crates/pypi_mapping/src/lib.rs

```diff
@@ -135,79 +83,270 @@ impl PurlSource {
+        let retry_policy = ExponentialBackoff::builder().build_with_max_retries(3);
+        let retry_strategy = RetryTransientMiddleware::new_with_policy(retry_policy);
+        let cache_strategy = Cache(HttpCache {
+            mode: CacheMode::Default,
+            manager: CACacheManager {
+                path: get_cache_dir()
+                    .expect("missing cache directory")
+                    .join(pixi_consts::consts::CONDA_PYPI_MAPPING_CACHE_DIR),
+            },
+            options: HttpCacheOptions::default(),
+        });
+
+        let client = ClientBuilder::from_client(client)
+            .with(cache_strategy)
+            .with(retry_strategy)
+            .build();
+
+        MappingClientBuilder {
+            client: client.clone(),
+            compressed_mapping: prefix::CompressedMappingClient::builder(client.clone()),
+            hash_mapping: prefix::HashMappingClient::builder(client),
+        }
+    }
+
+    /// Given a set of `RepoDataRecord`s, amend the purls for each record.
+    pub async fn ament_purls(
+        &self,
+        mapping_source: &MappingSource,
+        conda_packages: impl IntoIterator<Item = &mut RepoDataRecord>,
+        reporter: Option<Arc<dyn Reporter>>,
+    ) -> miette::Result<()> {
+        // Collect the records into a vec so we can iterate multiple times.
+        let mut records = conda_packages.into_iter().collect_vec();
+
+        // Normalize the channel names by removing the trailing slash
+        for package in records.iter_mut() {
+            package.channel = package
+                .channel
+                .as_ref()
+                .map(|c| c.trim_end_matches('/').to_string());
+        }
+
+        // Discard all records for which we already have pypi purls.
+        records.retain(|record| !has_pypi_purl(record));
+
+        // Fetch custom mapped channels if any.
+        let custom_mappings = if let MappingSource::Custom(mapping_url) = mapping_source {
+            CustomMappingClient::from(mapping_url.fetch_custom_mapping(&self.client).await?)
+        } else {
+            CustomMappingClient::default()
+        };
+
+        let mut amend_futures = FuturesUnordered::new();
+        let total_records = records.len();
+        for record in records.into_iter() {
+            let reporter = reporter.clone();
+            let custom_mappings = &custom_mappings;
+            let derive_purls_future = async move {
+                if let Some(reporter) = reporter.as_deref() {
+                    reporter.download_started(record, total_records);
+                }
+
+                let derived_purls = if matches!(mapping_source, MappingSource::Disabled) {
+                    Ok(None)
+                } else if custom_mappings.is_mapping_for_record(record) {
+                    custom_mappings.derive_purls(record).await
+                } else {
+                    self.derive_purls_from_clients(record).await
```

Comment:
could we implement `DerivePurls` for MappingClinet itself? and inside it will call both `hash_mapping` or `compressed`

Reply:
To me that is a bit confusing because `derive_purls_from_clients` only uses the `hash_mapping` and `compressed` mapping but not for the verbatim conda-forge mapping.